### ISSUE TITLE
There is a problem when I use LSNN

### ISSUE THREAD

**Issue by guoweiyu** [norse/norse#163]
The model i used is like this:
model = SequentialState(
    norse.LSNNLayer(norse.LSNNCell,12, 32),
    norse.LSNNLayer(norse.LSNNCell,32, 64),
    norse.LSNNLayer(norse.LSNNCell,64, 64),
    norse.LSNNLayer(norse.LSNNCell,64, 32),
    norse.LSNNLayer(norse.LSNNCell,32, 10),
    nn.Flatten(),
    norse.Lift(torch.nn.Linear(2000, 256)),
    norse.Lift(torch.nn.Linear(256, 10))
)
        read_out,state=model(x.cuda(),state)
        optimizer.zero_grad()  
        loss.backward()

the error:one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.FloatTensor [10, 10]], which is output 0 of TBackward, is at version 1; expected version 0 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).

when i change the code like: read_out,state=model(x.cuda()). The code works

But I want state to be passed between multiple inputs. How could i fix this question.

**Comment by Jegp** [comment]
Hi @guoweiyu, thanks for writing us! Did you try to use the latest master branch? I think this was fixed with #127 
```bash
pip install git+https://github.com/norse/norse
```

**Comment by guoweiyu** [comment]
Thanks a lot for your reply!
But it is still not working.
I write a simple test for this questions.  I get the same error when i run it. Maybe you can try to run it. Thanks again.


import torch
import norse.torch as norse
from norse.torch.module.sequential import SequentialState
from torch import nn
model = SequentialState(
    norse.LSNNLayer(norse.LSNNCell,12, 32),
    norse.LSNNLayer(norse.LSNNCell,32, 64),
    norse.LSNNLayer(norse.LSNNCell,64, 64),
    norse.LSNNLayer(norse.LSNNCell,64, 32),
    norse.LSNNLayer(norse.LSNNCell,32, 10),
    nn.Flatten(),
    norse.Lift(torch.nn.Linear(2000, 256)),
    norse.Lift(torch.nn.Linear(256, 10))
)
optimizer = torch.optim.Adam(model.parameters(), lr=0.0001)
loss_func = nn.MSELoss()
data = torch.ones(3,16, 200, 12)
target=torch.ones(16,10)
state=None
for x in data:
    out,state=model(x,state)
    loss=loss_func(out,target)
    optimizer.zero_grad()  # clear gradients for this training step
    loss.backward()  # backpropagation, compute gradients
    optimizer.step()

    print(out.shape)

**Comment by cpehle** [comment]
Thank you for pointing this out. I've confirmed your observation and implemented a failing test based on the code you provided. I will look into this this evening.

**Comment by cpehle** [comment]
Ok I've investigated this a bit more and it turns out that you will have to detach the state after you have done the gradient update step. Right now this is a bit inconvenient, because LSNNState doesn't have a detach operation. I've now created a change in the associated pull request to address this.

**Comment by Jegp** [comment]
Thanks for the test @guoweiyu and the implementation @cpehle. This is really something we should add to a general *Layer abstraction, as described in #121

**Comment by cpehle** [comment]
I agree. Detaching the state is not something that should necessarily be exposed to the user, because it is a detail of PyTorch autograd engine. I'm not sure how PyTorch handles it for their recurrent neural networks, because on the other hand you don't want to detach the state variable unconditionally. One way of doing it is not to destructively update the list in the SequentialState implementation, but this has large memory implications.

**Comment by guoweiyu** [comment]
thanks a lot.    @cpehle   @Jegp   now it really works. I will test the new code on the sEMG dataset to see if the loss could reduce normally

**Comment by Jegp** [comment]
Great it works @guoweiyu. Although all credit goes to @cpehle for the implementation :-)

@cpehle: yeah, getting this right is really an important step forward. I think we should open a separate issue to really focus on studying the PyTorch approach and getting it right. Probably for 0.0.6.
